### PR TITLE
Update MSC3083 support per changes in the MSC.

### PR DESCRIPTION
**Pull request by clokep**
* Add a type field.
* Switch `space` -> `room_id`.

Complement changes: matrix-org/complement#129

### CODE REVIEW

**Comment by clokep** [comment]
This will have some more extensive changes as well, but I wanted to separate the mechanical ones out to a separate PR.

File: synapse/handlers/event_auth.py [hunk 1]
```diff
@@ -102,17 +107,17 @@ async def has_restricted_join_rules(
         join_rules_event = await self._store.get_event(join_rules_event_id)
         return join_rules_event.content.get("join_rule") == JoinRules.MSC3083_RESTRICTED
 
-    async def get_spaces_that_allow_join(
+    async def get_rooms_that_allow_join(
         self, state_ids: StateMap[str]
     ) -> Collection[str]:
         """
-        Generate a list of spaces which allow access to a room.
+        Generate a list of rooms which membership in allow access to a room.
```
Comment (richvdh) [inline synapse/handlers/event_auth.py:114]: I think this is a marginal improvement, ymmv.

```suggestion
        Generate a list of rooms in which membership allows access to a room.
```

File: synapse/handlers/event_auth.py [hunk 2]
```diff
@@ -102,17 +107,17 @@ async def has_restricted_join_rules(
...
 
         Args:
             state_ids: The state of the room as it currently is.
```
Comment (richvdh) [inline synapse/handlers/event_auth.py:117]:
```suggestion
            state_ids: The current state of the room the user wishes to join
```

File: synapse/handlers/event_auth.py [hunk 3]
```diff
@@ -102,17 +107,17 @@ async def has_restricted_join_rules(
...
 
         Returns:
-            A collection of spaces which provide membership to the room.
+            A collection of room IDs which provide membership to the room.
```
Comment (richvdh) [inline synapse/handlers/event_auth.py:120]:
```suggestion
            A collection of room IDs. Membership in any of the rooms in the list grants the ability to join the target room.
```

**Review by richvdh — APPROVED** [review]
lgtm otherwise

File: synapse/handlers/space_summary.py [hunk 4]
```diff
@@ -533,7 +533,7 @@ async def _build_room_entry(self, room_id: str) -> JsonDict:
             "guest_can_join": stats["guest_access"] == "can_join",
             "creation_ts": create_event.origin_server_ts,
             "room_type": create_event.content.get(EventContentFields.ROOM_TYPE),
-            "allowed_spaces": allowed_spaces,
+            "allowed_spaces": allowed_rooms,
```
Comment (clokep) [inline synapse/handlers/space_summary.py:536]: I should have mentioned that I left this at `allowed_spaces` for now since the form of this API is somewhat unrelated to the restricted rooms MSC.
